Some nanoarrow.hpp content is not compatible with gcc 4.8

Thankfully gcc 4.8 will be out of common usage soon; however, we in theory still support it and some content of nanoarrow.hpp results in compile errors there:

```
  Scanning dependencies of target nanoarrow_c_data_integration
  [ 25%] Building CXX object CMakeFiles/nanoarrow_c_data_integration.dir/src/nanoarrow/integration/c_data_integration.cc.o
  In file included from /nanoarrow/src/nanoarrow/nanoarrow_testing.hpp:27:0,
                   from /nanoarrow/src/nanoarrow/integration/c_data_integration.cc:24:
  /nanoarrow/src/nanoarrow/nanoarrow.hpp:578:17: error: ‘is_trivially_copyable’ is not a member of ‘std’
     static_assert(std::is_trivially_copyable<T>::value, "");
                   ^
  /nanoarrow/src/nanoarrow/nanoarrow.hpp:578:45: error: expected primary-expression before ‘>’ token
     static_assert(std::is_trivially_copyable<T>::value, "");
                                               ^
  /nanoarrow/src/nanoarrow/nanoarrow.hpp:578:46: error: ‘::value’ has not been declared
     static_assert(std::is_trivially_copyable<T>::value, "");
                                                ^
  /nanoarrow/src/nanoarrow/nanoarrow.hpp:908:24: error: missing space between ‘""’ and suffix identifier
   inline ArrowStringView operator""_v(const char* data, std::size_t size_bytes) {
                          ^
  /nanoarrow/src/nanoarrow/nanoarrow.hpp: In instantiation of ‘nanoarrow::internal::Maybe<T>::Maybe(T) [with T = ArrowStringView]’:
  /nanoarrow/src/nanoarrow/nanoarrow.hpp:778:65:   required from here
  /nanoarrow/src/nanoarrow/nanoarrow.hpp:563:50: error: cannot convert ‘ArrowStringView’ to ‘const char*’ in initialization
         : something_{something}, is_something_{true} {}
                                                    ^
  /nanoarrow/src/nanoarrow/nanoarrow.hpp: In instantiation of ‘nanoarrow::internal::InputRange<Next>::iterator nanoarrow::internal::InputRange<Next>::end() [with Next = nanoarrow::ViewArrayStream::Next]’:
  /nanoarrow/src/nanoarrow/nanoarrow.hpp:877:38:   required from here
  /nanoarrow/src/nanoarrow/nanoarrow.hpp:634:36: error: could not convert ‘{this, <brace-enclosed initializer list>()}’ from ‘<brace-enclosed initializer list>’ to ‘nanoarrow::internal::InputRange<nanoarrow::ViewArrayStream::Next>::iterator’
     iterator end() { return {this, {}}; }
                                      ^
  /nanoarrow/src/nanoarrow/nanoarrow.hpp: In instantiation of ‘nanoarrow::internal::Maybe<T>::Maybe() [with T = ArrowStringView]’:
  /nanoarrow/src/nanoarrow/nanoarrow.hpp:560:26:   required from ‘nanoarrow::internal::Maybe<T>::Maybe(nanoarrow::internal::Nothing) [with T = ArrowStringView]’
  /nanoarrow/src/nanoarrow/nanoarrow.hpp:780:14:   required from here
  /nanoarrow/src/nanoarrow/nanoarrow.hpp:559:53: error: too many initializers for ‘nanoarrow::internal::Nothing’
     Maybe() : nothing_{Nothing{}}, is_something_{false} {}
                                                       ^
  gmake[2]: *** [CMakeFiles/nanoarrow_c_data_integration.dir/src/nanoarrow/integration/c_data_integration.cc.o] Error 1
  gmake[1]: *** [CMakeFiles/nanoarrow_c_data_integration.dir/all] Error 2
  gmake: *** [all] Error 2
  Failed to verify release candidate. See /tmp/nanoarrow-HEAD.YCAh0T for details.
```

--- a/src/nanoarrow/nanoarrow.hpp
+++ b/src/nanoarrow/nanoarrow.hpp
@@ -906,11 +906,7 @@ inline bool operator==(ArrowStringView l, ArrowStringView r) {
 }
 
 /// \brief User literal operator allowing ArrowStringView construction like "str"_sv
-inline ArrowStringView
-// clang-format off
-operator ""
-    // clang-format on
-    _v(const char* data, std::size_t size_bytes) {
+inline ArrowStringView operator"" _v(const char* data, std::size_t size_bytes) {
   return {data, static_cast<int64_t>(size_bytes)};
 }
 /// @}

--- a/src/nanoarrow/nanoarrow.hpp
+++ b/src/nanoarrow/nanoarrow.hpp
@@ -556,11 +556,11 @@ struct Nothing {};
 template <typename T>
 class Maybe {
  public:
-  Maybe() : nothing_{Nothing{}}, is_something_{false} {}
-  Maybe(Nothing) : Maybe{} {}
+  Maybe() : nothing_(Nothing()), is_something_(false) {}
+  Maybe(Nothing) : Maybe() {}
 
   Maybe(T something)  // NOLINT(google-explicit-constructor)
-      : something_{something}, is_something_{true} {}
+      : something_(something), is_something_(true) {}
 
   explicit constexpr operator bool() const { return is_something_; }
 
@@ -575,7 +575,8 @@ class Maybe {
   T value_or(T val) const { return is_something_ ? something_ : val; }
 
  private:
-  static_assert(std::is_trivially_copyable<T>::value, "");
+  // When support for gcc 4.8 is dropped, we should also assert
+  // is_trivially_copyable<T>::value
   static_assert(std::is_trivially_destructible<T>::value, "");
 
   union {
@@ -631,7 +632,7 @@ struct InputRange {
   };
 
   iterator begin() { return {this, next()}; }
-  iterator end() { return {this, {}}; }
+  iterator end() { return {this, ValueOrFalsy()}; }
 };
 }  // namespace internal
 
@@ -825,7 +826,7 @@ class ViewArrayAsFixedSizeBytes {
 class ViewArrayStream {
  public:
   ViewArrayStream(ArrowArrayStream* stream, ArrowErrorCode* code, ArrowError* error)
-      : range_{Next{this, stream, UniqueArray{}}}, code_{code}, error_{error} {}
+      : range_{Next{this, stream, UniqueArray()}}, code_{code}, error_{error} {}
 
   ViewArrayStream(ArrowArrayStream* stream, ArrowError* error)
       : ViewArrayStream{stream, &internal_code_, error} {}
@@ -905,7 +906,11 @@ inline bool operator==(ArrowStringView l, ArrowStringView r) {
 }
 
 /// \brief User literal operator allowing ArrowStringView construction like "str"_sv
-inline ArrowStringView operator""_v(const char* data, std::size_t size_bytes) {
+inline ArrowStringView
+// clang-format off
+operator ""
+    // clang-format on
+    _v(const char* data, std::size_t size_bytes) {
   return {data, static_cast<int64_t>(size_bytes)};
 }
 /// @}

--- a/src/nanoarrow/nanoarrow_hpp_test.cc
+++ b/src/nanoarrow/nanoarrow_hpp_test.cc
@@ -271,7 +271,7 @@ TEST(NanoarrowHppTest, NanoarrowHppViewArrayAsTest) {
   struct ArrowArray array {};
   array.length = 7;
   array.null_count = 2;
-  array.n_buffers = std::size(buffers);
+  array.n_buffers = 2;
   array.buffers = buffers;
 
   int i = 0;
@@ -300,7 +300,7 @@ TEST(NanoarrowHppTest, NanoarrowHppViewArrayAsBytesTest) {
   struct ArrowArray array {};
   array.length = 7;
   array.null_count = 2;
-  array.n_buffers = std::size(buffers);
+  array.n_buffers = 2;
   array.buffers = buffers;
 
   int i = 0;
@@ -327,7 +327,7 @@ TEST(NanoarrowHppTest, NanoarrowHppViewArrayAsFixedSizeBytesTest) {
   struct ArrowArray array {};
   array.length = 7;
   array.null_count = 2;
-  array.n_buffers = std::size(buffers);
+  array.n_buffers = 2;
   array.buffers = buffers;
 
   int i = 0;

